Added cosmos adapter

dev-cosmos opened the pull request:
<!--
Thank you for your pull request. Please make sure this PR is scoped to one change, and that any added or changed code includes tests with greater than 80% code coverage. See https://github.com/prebid/Prebid.js/blob/master/CONTRIBUTING.md#testing-prebidjs for documentation on testing Prebid.js.
-->

## Type of change
<!-- Remove items that don't apply and/or select an item by changing [ ] to [x] -->
- [ ] Bugfix
- [ ] Feature
- [x] New bidder adapter  <!--  IMPORTANT: if checking here, also submit your bidder params documentation here https://github.com/prebid/prebid.github.io/tree/master/dev-docs/bidders --> 
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Does this change affect user-facing APIs or examples documented on http://prebid.org?
- [ ] Other

## Description of change
<!-- Describe the change proposed in this pull request -->

<!-- For new bidder adapters, please provide the following -->
- test parameters for validating bids
```
{
  bidder: '<bidder name>',
  params: {
    // ...
  }
}
```

Be sure to test the integration with your adserver using the [Hello World](/integrationExamples/gpt/hello_world.html) sample page.

- contact email of the adapter’s maintainer
- [x] official adapter submission

For any changes that affect user-facing APIs or example code documented on http://prebid.org, please provide:

- A link to a PR on the docs repo at https://github.com/prebid/prebid.github.io/

## Other information
<!-- References to related PR or issue #s, @mentions of the person or team responsible for reviewing changes, etc. -->

jsnellbaker commented on modules/cosmosBidAdapter.js:
```diff
@@ -0,0 +1,398 @@
+  _copyObject: function (src, dst) {
+    if (src) {
+      // copy complete object
+      Object.keys(src).forEach(param => dst[param] = src[param]);
```
I believe this type of copying could leave a reference of the original object in the copied object, which could mean if the original object was modified  - the copied object would mutate and use the new value(s).

There is a `utils.deepClone()` function available if you want to safely copy one object to another.

dev-cosmos replied:
Used utils.deepClone() in buildRequests function to clone `validBidRequests` input params from pbs lib. Need this function to give object preference as per cosmos logic.

jsnellbaker commented on modules/cosmosBidAdapter.js:
```diff
@@ -0,0 +1,398 @@
+      spec._copyObject(rSite, site);
+      site.domain = utils.getTopWindowLocation().hostname;
+      site.ref = utils.getTopWindowReferrer();
+      site.page = utils.getTopWindowLocation().href;
```
These three `getTopWindow...` functions are marked to be deprecated and may not be available in the next major release.  A `refererInfo` object is available in the `bidderRequest` param that should have values available you can use instead.  

The following link provides additional information:
http://prebid.org/dev-docs/bidder-adaptor.html#referrers

Can you please update these values accordingly?

dev-cosmos replied:
Thanks for input and saving our future efforts. Updated above code with refererInfo object.

jsnellbaker commented on modules/cosmosBidAdapter.js:
```diff
@@ -0,0 +1,398 @@
+      // video bid request validation
+      if (bid.params.hasOwnProperty('video')) {
+        if (!bid.params.video.hasOwnProperty('mimes') ||
+          !utils.isArray(bid.params.video.mimes) ||
```
Is this the correct property to check for the `mimes` property?  

From looking at the sample video adUnit in the corresponding md file, the `mimes` property is listed under the `mediaTypes.video` property, not under the `bids[].params.video` property.  If the md file is incorrect and this field should be checked at this location, can you please update the md file?

dev-cosmos replied:
Thanks for good catch. corrected all the checks.

jsnellbaker commented on modules/cosmosBidAdapter.js:
```diff
@@ -0,0 +1,398 @@
+                  } else {
+                    bidResponse.width = bid.hasOwnProperty('w') ? bid.w : impr.video.w;
+                    bidResponse.height = bid.hasOwnProperty('h') ? bid.h : impr.video.h;
+                    bidResponse.vastXml = bid.adm;
```
Please add a line somewhere in this `else` block to specify the following property:
`bidResponse.mediaType = 'video';`

This will properly flag the bid as a `video` bid, which is needed for some logic later in the auction process.

dev-cosmos replied:
Thanks for input. Now we set `mediaType` for both banner and video response.

jsnellbaker commented:
This part got left out of my previous comments for some reason, but there are unit tests failures happening IE 11.  Can you please take a look and fix the issue?

The circleCI link provides the job details, but I copied them here for convenience:
```
  Cosmos adapter
    implementation
      Request formation
        ✗ Endpoint checking
	TypeError: Unable to get property 'url' of undefined or null reference
	   at Anonymous function (webpack:///test/spec/modules/cosmosBidAdapter_spec.js:194:8 <- test/test_index.js:81704:9)

        ✗ Request params check
	TypeError: Unable to get property 'data' of undefined or null reference
	   at Anonymous function (webpack:///test/spec/modules/cosmosBidAdapter_spec.js:200:8 <- test/test_index.js:81709:9)

        ✗ Request params check: without tagId
	TypeError: Unable to get property 'data' of undefined or null reference
	   at Anonymous function (webpack:///test/spec/modules/cosmosBidAdapter_spec.js:210:8 <- test/test_index.js:81721:9)

        ✗ Request params multi size format object check
	TypeError: Unable to get property 'data' of undefined or null reference
	   at Anonymous function (webpack:///test/spec/modules/cosmosBidAdapter_spec.js:236:8 <- test/test_index.js:81748:9)

        ✗ Request params currency check
	TypeError: Unable to get property 'data' of undefined or null reference
	   at Anonymous function (webpack:///test/spec/modules/cosmosBidAdapter_spec.js:296:8 <- test/test_index.js:81809:9)

        ✗ Request params check for video ad
	TypeError: Unable to get property 'data' of undefined or null reference
	   at Anonymous function (webpack:///test/spec/modules/cosmosBidAdapter_spec.js:312:8 <- test/test_index.js:81823:9)

        Response checking
          ✗ should check for valid response values
	TypeError: Unable to get property 'data' of undefined or null reference
	   at Anonymous function (webpack:///test/spec/modules/cosmosBidAdapter_spec.js:327:10 <- test/test_index.js:81837:11)
```

bretg commented:
Docs PR https://github.com/prebid/prebid.github.io/pull/1435

dev-cosmos commented on modules/cosmosBidAdapter.js:
```diff
@@ -0,0 +1,398 @@
+   **/
+  isBidRequestValid: function (bid) {
+    var retval = false;
+    do {
```
removed do-while loop as recommended.

dev-cosmos commented on modules/cosmosBidAdapter.js:
```diff
@@ -0,0 +1,398 @@
+   * @param {Data} Containing timeout specific data
+   */
+  onTimeout: function (data) {
+  },
```
removed function.

dev-cosmos commented on modules/cosmosBidAdapter.js:
```diff
@@ -0,0 +1,398 @@
+   * @param {Bid} The bid of which the targeting has been set
+   **/
+  onSetTargeting: function (bid) {
+  },
```
removed function.

dev-cosmos commented on modules/cosmosBidAdapter.js:
```diff
@@ -0,0 +1,398 @@
+   * @param {Bid} The bid that won the auction
+   */
+  onBidWon: function (bid) {
+  },
```
removed function.

dev-cosmos commented:
Thank you @jsnellbaker for review and all recommendation. We address all issue highlighted by you including `IE 11` and user `usersync.html:4:24`.

We also tested cosmos adapter using hello world page and checked-in `cosmos_sample.html` into integrationExamples/gpt/ for reference.

dev-cosmos commented:
> Hello @dev-cosmos
> 
> Thanks for submitting this new adapter to the project. I reviewed the code and tested out the adapter. There are a few items that need to be reviewed/addressed in terms of the code, but the testing mainly looked good (see now on userSync test below).
> 
> Please let me know if you have any questions on the feedback. Thanks!
> 
> UserSync Pixel issue:
> When I was attempting to test the iframe userSync pixel, I observed the following error was thrown in the browser console. Can you take a look?
> 
> ```
> usersync.html:4 Uncaught DOMException: Blocked a frame with origin "http://sync.cosmoshq.com" from accessing a cross-origin frame.
>     at http://sync.cosmoshq.com/js/v1/usersync.html:4:24
> ```

Thanks for catching. We fixed it.

jsnellbaker commented:
@dev-cosmos Did you mean to close this PR?  Please confirm.  Thanks!

dev-cosmos commented:
Hi @jsnellbaker: can you please review PR #4096. By mistake we deleted forked repository of this PR, so we created new PR with number #4096 . Addressed all review comment's suggested by you.

jsnellbaker commented:
Thanks for the clarification; will grab this new PR and take the conversation there.  Thanks.
